presto-main/src/main/java/com/facebook/presto/execution/scheduler/group/DynamicLifespanScheduler.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.scheduler.group;
+
+import com.facebook.presto.execution.Lifespan;
+import com.facebook.presto.execution.scheduler.BucketNodeMap;
+import com.facebook.presto.execution.scheduler.SourceScheduler;
+import com.facebook.presto.spi.Node;
+import com.facebook.presto.spi.connector.ConnectorPartitionHandle;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.SettableFuture;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntListIterator;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.OptionalInt;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class DynamicLifespanScheduler
+        implements LifespanScheduler
+{
+    private final BucketNodeMap bucketNodeMap;
+    private final List<Node> allNodes;
+    private final List<ConnectorPartitionHandle> partitionHandles;
+    private final OptionalInt concurrentLifespansPerTask;
+
+    private final IntListIterator driverGroups;
+
+    private boolean initialScheduled;
```

Comment:
I believe these needs to be guarded

Reply:
Please add back the comments about thread safety (which you unintentionally removed). They should be added to `LifespanScheduler` interface. And you can leave a pointer here.

```
        // Thread Safety:
        // * Invocation of onLifespanFinished can be parallel and in any thread.
        //   There may be multiple invocations in flight at the same time,
        //   and may overlap with any other methods.
        // * Invocation of schedule happens sequentially in a single thread.
        // * This object is safely published after invoking scheduleInitial.
```

Please also add specific comments here:
```
* initialScheduled does not need to be guarded because this object
  is safely published after its mutation.
* Write to newDriverGroupReady field is guarded. Read of the reference
  is either guarded, or is guaranteed to happen in the same thread as the write.